Document common selectors for the Recorder (#2424)
GoogleChrome/developer.chrome.com: site/en/docs/devtools/recorder/index.md

## Record a user flow {: #record }

1. Open this [demo](https://coffee-cart.netlify.app/) page. Click on the **Start new recording** button to begin.
1. Enter "coffee checkout" in the **Recording name** textbox.
   {% Img src="image/NJdAV9UgKuN8AhoaPBquL7giZQo1/EnRcThJpdtY6c91ZTRMT.png", alt="Start a new recording", width="800", height="520" %}
   Notice the default `data-testid` value in the **Selector attribute** textbox and leave it as it is for the purposes of this tutorial.
   {% Aside %}

>    If you use a custom selector attribute in your code, this is where you can specify it. For more information, see [Customize the recording's selector](#customize-selector).

Reviewer: I think some users might it surprise that it detected `data-testid`. instead of `id` or CSS class. 

Explain and reasoning: 
- Suggest to explain in the `aside` (or add one more aside) to explain that `data-testid` is optional, it will work fine with id or class.
- Suggest to explain the reasoning on the `data-testid` usage. 

Can reference to https://docs.cypress.io/guides/references/best-practices#Selecting-Elements and the `Did you know` section.

## Record a user flow {: #record }

1. Open this [demo](https://coffee-cart.netlify.app/) page. Click on the **Start new recording** button to begin.
1. Enter "coffee checkout" in the **Recording name** textbox.
   {% Img src="image/NJdAV9UgKuN8AhoaPBquL7giZQo1/EnRcThJpdtY6c91ZTRMT.png", alt="Start a new recording", width="800", height="520" %}

>    Notice the default `data-testid` value in the **Selector attribute** textbox and leave it as it is for the purposes of this tutorial.

Reviewer: ```suggestion
```

You can also edit the recording's name by clicking the edit button <span class="material-icons">edit</span> next to it.

## Edit steps {: #edit-steps }

Let's walk through the options to edit the steps within a workflow.

### Expand steps {: #expand-step }

1. Expand each step to see the details of the action. For example, expand the *Click Element "Cappuccino"* step.
  {% Img src="image/NJdAV9UgKuN8AhoaPBquL7giZQo1/WUIZb8eMWfcPZHPyQ5C5.png", alt="In the recorder panel, the Cappuccino element has been expanded to reveal type, target, selectors, offset X, and offset Y.", width="800", height="773" %}
1. The step above shows two **selectors**. For more information, see [Selector types](#selector-types).
   When replaying the user flow, the **Recorder** tries to query the element with one of the selectors by sequence.
   For example, if the **Recorder** successfully queries the element with the first selector, it will skip the second selector and proceed to the next step.
1. You can add or remove any selectors. For example, you can remove the *selector #2* because just `aria/Cappuccino` is sufficient in this case. Hover over the *selector #2* and click on **-** to remove it.

>     {% Img src="image/dPDCek3EhZgLQPGtEG3y0fTn4v82/qEat3jT1Og68urhmSkef.png", alt="The DevTools recorder panel shows an option to remove a selector", width="800", height="663" %}

Reviewer: outdated image because the second selector is not using [data-test]. Mismaatch from the previous step

1. The **selector** is editable too. For example, if you want to select *Mocha* instead of *Cappuccino*, you can edit the selector value to *aria/Mocha* instead.
  {% Img src="image/dPDCek3EhZgLQPGtEG3y0fTn4v82/OdihJ4BROOXZ93eidr4b.png", alt="edit a selector", width="800", height="663" %} 
1. [Replay](#replay) the flow now, it should select *Mocha* instead of *Cappuccino*.

> 1. Try to edit other step properties such as **type**, **target**, **value** and more.

Reviewer: There are two ways of editing the value, user can

a. edit the Mocha, or
b. use the selector picker to pick the selector on the screen. (I demo this in the video but not in this doc.)

Please add a step to explain b)

#### Selector types {: #selector-types }

During recording, the **Recorder** automatically detects two types of selectors for most of the steps: ARIA and CSS.

{% Aside %}
For more information on ARIA selectors, see [Syntactic vs. semantic selectors](/blog/puppetaria/#syntactic-vs-semantic-selectors).
{% endAside %}

In addition to the ARIA selector, the **Recorder** looks for the best CSS selector it can find by the following attributes and in the following order:
1. Your custom selector attribute if you specified it at the start of the recording.
   If found, the **Recorder** will prioritize your custom CSS attribute selector above the ARIA one.
   For more information, see [Customize the recording's selector](#customize-selector).
1. The most common attributes used for testing: {: #selectors }
   - `data-testid`
   - `data-test`
   - `data-qa`
   - `data-cy`
   - `data-test-id`
   - `data-qa-id`
   - `data-testing`

> 1. Regular CSS selectors.

Reviewer: 1. `id` attribute
1. regular css selecors

1. You can add or remove any selectors. For example, you can remove the *selector #2* because just `aria/Cappuccino` is sufficient in this case. Hover over the *selector #2* and click on **-** to remove it.
    {% Img src="image/NJdAV9UgKuN8AhoaPBquL7giZQo1/ivZru7CFneS6fu5BAcIJ.png", alt="The DevTools recorder panel shows an option to remove a selector", width="800", height="758" %}
1. The **selector** is editable too. For example, if you want to select *Mocha* instead of *Cappuccino*, you can edit the selector value to *aria/Mocha* instead.
  {% Img src="image/dPDCek3EhZgLQPGtEG3y0fTn4v82/OdihJ4BROOXZ93eidr4b.png", alt="edit a selector", width="800", height="663" %}
   Alternatively, click the **Select**{% Img src="image/NJdAV9UgKuN8AhoaPBquL7giZQo1/ihbO22H4EA0HfrNX6gV1.png", alt="Select button", width="24", height="22" %} button and then click *Mocha* on the page.
1. [Replay](#replay) the flow now, it should select *Mocha* instead of *Cappuccino*.
1. Try to edit other step properties such as **type**, **target**, **value** and more.

> #### Selector types {: #selector-types }

Reviewer: How about move this section to the bottom, create a ## Selectors section, with both this and  the customize the recording's selector section?

1. Both of this section are currently too far away from each other, ideally we put them together.
2. Selector isn't belong the edit steps section

    - count: **9** (click **add count** button)
   {% Img src="image/dPDCek3EhZgLQPGtEG3y0fTn4v82/zeK91O21jvc5OD9HuovR.png", alt="The new step for coffee checkout has been updated with the aforementioned details.", width="800", height="663" %}
6. [Replay](#replay) the flow now to see the changes.

## Understanding the recording's selector {: #selector }

During recording, the **Recorder** automatically detects two types of selectors for most of the steps: ARIA and CSS.

{% Aside %}
For more information on ARIA selectors, see [Syntactic vs. semantic selectors](/blog/puppetaria/#syntactic-vs-semantic-selectors).
{% endAside %}

> For simple webpages, `id` attributes and CSS `class` attributes are sufficient for the **Recorder** to detect the selectors. However, that might not always be the case.

Reviewer: ```suggestion
For simple webpages, `id` attributes and CSS `class` attributes are sufficient for the **Recorder** to detect the selectors. However, that might not always be the case, because:

- Your webpages may use dynamic classes or ID's that change
- Your selectors may break from development changes to CSS styles or JS behaviour
```

### Common test selectors {: common-test-selector}

> For applications developed with modern JavaScript frameworks, (for example, [React](https://reactjs.org/), [Angular](https://angular.io/), [Vue](https://vuejs.org/), and CSS frameworks), the CSS `class` values might be autogenerated. For example:

Reviewer: ```suggestion
For example, the CSS `class` values might be autogenerated for applications developed with modern JavaScript frameworks (e.g. [React](https://reactjs.org/), [Angular](https://angular.io/), [Vue](https://vuejs.org/)) and CSS frameworks.
```

{% Img src="image/NJdAV9UgKuN8AhoaPBquL7giZQo1/ZtK52PaMMzKWiiAcsQfH.png", alt="Autogenerated CSS classes with randomized names", width="800", height="654" %}

> There are already some common popular test selectors in the market and the **Recorder** supports them as well. If you have the following common test selectors defined, the **Recorder** will automatically detect and use them first:

Reviewer: ```suggestion
In these cases, you can use `data-*` attributes to create more resilient tests. There are already some common `data-*` selectors  which people use for automation. 

The **Recorder** supports them as well. If you have the following common test selectors defined, the **Recorder** will automatically detect and use them first:
```

- `data-testid`
- `data-test`
- `data-qa`
- `data-cy`
- `data-test-id`
- `data-qa-id`
- `data-testing`

> 

Author: Done

> To create more resilient tests, consider using these test selectors.

Reviewer: ```suggestion

```

> 

Reviewer: Should we provide a screenshot or animated example here?

Author: What do you want to show here animated?

I think even a text example can do. I'll add it.

    - count: **9** (click **add count** button)
   {% Img src="image/dPDCek3EhZgLQPGtEG3y0fTn4v82/zeK91O21jvc5OD9HuovR.png", alt="The new step for coffee checkout has been updated with the aforementioned details.", width="800", height="663" %}
6. [Replay](#replay) the flow now to see the changes.

> 

Reviewer: I swap the content around to first explain the usual selectors,  the common test selectors (why we need it), then talk about customizing it. 

IMHO, the reasoning is important for the user know why we need that.

I did not proofread my sentence, so please just take the meaning and rewrite / reorganize them. :)

## Customize the recording's selector {: #customize-selector }

> You can customize the selector of a recording if you are not using the [common selectors](#selectors).

Reviewer: I want to say if both the id and css selectors + the common test selectors don't works for you, then this is the 3rd option.

For example, this [demo page](https://jec.fyi/demo/recorder) uses the `data-automate` attribute as the selector. [Start a new recording](/docs/devtools/recorder/#record) and enter the `data-automate` as the selector attribute. Fill in an email address and observe the selector value (`[data-automate=email-address]`).

{% Img src="image/dPDCek3EhZgLQPGtEG3y0fTn4v82/2PPPt9tOC2ZEz1l9F9AK.png", alt="Customize the recording's selector", width="800", height="524" %}

{% Img src="image/dPDCek3EhZgLQPGtEG3y0fTn4v82/X8r52vWEu6aC8QHFuknp.png", alt="The result of custom selector selection", width="800", height="579" %}

### Selector types {: #selector-types }

> During recording, the **Recorder** automatically detects two types of selectors for most of the steps: ARIA and CSS.

Reviewer: Revise the section here (some content might have been moved to the top already)
